### PR TITLE
Plugins: Externalize Cloud Monitoring data source

### LINKED ISSUE #79355: GCM: Externalization pt. 2 Package Decouple

### PR DESCRIPTION
**Pull request by alyssajoyner**
This PR completes part 2 of the externalization of the Cloud Monitoring data source.

Closes #79355

### CODE REVIEW

**Review by mckn — COMMENTED** [review]
LGTM! Someone more familiar with cloud monitoring should approve this.

File: pkg/tsdb/cloud-monitoring/cloudmonitoring.go [hunk 1]
```diff
@@ -107,11 +100,6 @@ func (s *Service) CheckHealth(ctx context.Context, req *backend.CheckHealthReque
 	if err != nil {
 		return nil, err
 	}
-	defer func() {
-		if err := res.Body.Close(); err != nil {
```
Comment (andresmgot) [inline pkg/tsdb/cloud-monitoring/cloudmonitoring.go:111]: this `Body` is no longer closed?

File: pkg/tsdb/cloud-monitoring/cloudmonitoring.go [hunk 2]
```diff
@@ -595,27 +577,27 @@ func (s *Service) getDefaultProject(ctx context.Context, dsInfo datasourceInfo)
...
+func unmarshalResponse(res *http.Response) (cloudMonitoringResponse, error) {
 	body, err := io.ReadAll(res.Body)
 	if err != nil {
 		return cloudMonitoringResponse{}, err
 	}
 
 	defer func() {
 		if err := res.Body.Close(); err != nil {
-			logger.Warn("Failed to close response body", "err", err)
+			backend.Logger.Warn("Failed to close response body", "err", err)
```
Comment (andresmgot) [inline pkg/tsdb/cloud-monitoring/cloudmonitoring.go:588]: The logger before was populated with the `context` information. `backend.Logger` is just the default logger. So I think you need to:

- Instantiate the logger as done with testdata: https://github.com/grafana/grafana/blob/main/pkg/tsdb/grafana-testdata-datasource/testdata.go#L34 
- Then, in each method, use `s.logger.FromContext` to get a custom logger with contextual info (as it was done before).

After that, if you compare log entries, they should be similar before and after the change.

File: pkg/tsdb/cloud-monitoring/httpclient.go [hunk 3]
```diff
@@ -59,7 +59,13 @@ func getMiddleware(model *datasourceInfo, routePath string) (httpclient.Middlewa
 	return tokenprovider.AuthMiddleware(provider), nil
 }
 
-func newHTTPClient(model *datasourceInfo, opts httpclient.Options, clientProvider infrahttp.Provider, route string) (*http.Client, error) {
+type Provider interface {
```
Comment (andresmgot) [inline pkg/tsdb/cloud-monitoring/httpclient.go:62]: Can you use the SDK http provider for this as Azure does? https://github.com/grafana/grafana/blob/main/pkg/tsdb/azuremonitor/azuremonitor.go/#L28
If not, we should move this interface definition to the SDK rather than copying it here.

File: pkg/tsdb/cloud-monitoring/slo_query.go [hunk 4]
```diff
@@ -7,18 +7,16 @@ import (
...
-	s *Service, dsInfo datasourceInfo, tracer tracing.Tracer) (*backend.DataResponse, any, string, error) {
-	return runTimeSeriesRequest(ctx, sloQ.logger, req, s, dsInfo, tracer, sloQ.parameters.ProjectName, sloQ.params, nil)
+	s *Service, dsInfo datasourceInfo) (*backend.DataResponse, any, string, error) {
+	return runTimeSeriesRequest(ctx, req, s, dsInfo, sloQ.parameters.ProjectName, sloQ.params, nil)
 }
 
 func (sloQ *cloudMonitoringSLO) parseResponse(queryRes *backend.DataResponse,
 	response any, executedQueryString string) error {
-	return parseTimeSeriesResponse(queryRes, response.(cloudMonitoringResponse), executedQueryString, sloQ, sloQ.params, []string{})
+	return parseTimeSeriesResponse(queryRes, response.(cloudMonitoringResponse), executedQueryString, nil, sloQ.params, []string{})
```
Comment (andresmgot) [inline pkg/tsdb/cloud-monitoring/slo_query.go:19]: this change seems unrelated? `sloQ` -> `nil`

File: public/app/plugins/datasource/cloud-monitoring/package.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{
+  "name": "@grafana-plugins/grafana-cloud-monitoring-datasource",
+  "description": "Grafana data source for Google Cloud Monitoring",
+  "private": true,
+  "version": "10.3.0-pre",
```
Comment (andresmgot) [inline public/app/plugins/datasource/cloud-monitoring/package.json:5]: 10.3.0-pre needs to be replaced with 10.4.0-pre

**Review by andresmgot — COMMENTED** [review]
Good job! I have some comments about things I've spotted.

File: pkg/tsdb/cloud-monitoring/cloudmonitoring.go [hunk 6]
```diff
@@ -65,11 +59,10 @@ const (
 	perSeriesAlignerDefault   = "ALIGN_MEAN"
 )
 
-func ProvideService(httpClientProvider httpclient.Provider, tracer tracing.Tracer) *Service {
```
Comment (aangelisc) [inline pkg/tsdb/cloud-monitoring/cloudmonitoring.go:68]: Instead of removing this I think we should use the tracer from the SDK? I think that's applicable? cc @andresmgot

Comment (andresmgot) [reply pkg/tsdb/cloud-monitoring/cloudmonitoring.go:68]: I think that the tracer is still used: [link](https://github.com/grafana/grafana/pull/80181/files#diff-e0d8854117661da605b3d4b545db968683557373389b57e2a8143ce13ef487e3R128). We probably don't need to store it in the service (I'm no tracer expert :D)

Comment (xnyo) [reply pkg/tsdb/cloud-monitoring/cloudmonitoring.go:68]: That's correct, `backend.DefaultTracer()` is the tracer provided by the SDK and the dependency from the tracer in `pkg/infra/tracing` can be removed

File: pkg/tsdb/cloud-monitoring/cloudmonitoring.go (hunk 2 above)
Comment (andresmgot) [reply pkg/tsdb/cloud-monitoring/cloudmonitoring.go:596]: This is still not fully solved, you need to replace all `backend.Logger` appearances and `QueryData` was using `.FromContext` before so we should probably keep it.

Comment (alyssabull) [reply pkg/tsdb/cloud-monitoring/cloudmonitoring.go]: Thanks @andresmgot. I replaced `backend.Logger` and added back in the context as well as updated the corresponding tests.

File: public/app/plugins/datasource/cloud-monitoring/package.json [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    "@testing-library/react": "14.0.0",
+    "@testing-library/user-event": "14.5.1",
+    "@types/debounce-promise": "3.1.6",
+    "@types/jest": "29.5.4",
+    "@types/lodash": "4.14.195",
+    "@types/node": "20.8.10",
+    "@types/prismjs": "1.26.0",
+    "@types/react": "18.2.15",
+    "@types/react-test-renderer": "18.0.0",
+    "@types/testing-library__jest-dom": "5.14.8",
```
Comment (bossinc) [inline public/app/plugins/datasource/cloud-monitoring/package.json:39]:
```suggestion
    "@types/react": "18.2.48",
    "@types/testing-library__jest-dom": "5.14.9",
```

**Review by bossinc — APPROVED** [review]
